exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1786 @@
+}
+
+func (ku *Kucoin) processTicker(respData []byte, instrument string) error {
+	response := WsTicker{}
+	err := json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	spotTickerPrice := ticker.Price{
+		AssetType:    asset.Spot,
+		Last:         response.Price,
+		LastUpdated:  response.Timestamp.Time(),
+		ExchangeName: ku.Name,
+		Pair:         pair,
+		Ask:          response.BestAsk,
+		Bid:          response.BestBid,
+		AskSize:      response.BestAskSize,
+		BidSize:      response.BestBidSize,
+		Volume:       response.Size,
+	}
+	ku.Websocket.DataHandler <- &spotTickerPrice
+	marginTickerPrice := spotTickerPrice
+	marginTickerPrice.AssetType = asset.Margin
+	ku.Websocket.DataHandler <- &marginTickerPrice
+	return nil
+}
+
+func (ku *Kucoin) processCandlesticks(respData []byte, instrument, intervalString string) error {
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	response := WsCandlestickData{}
+	err = json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	resp, err := response.getCandlestickData()
+	if err != nil {
+		return err
+	}
+	candlestickData := stream.KlineData{
+		Timestamp:  response.Time.Time(),
+		Pair:       pair,
+		AssetType:  asset.Spot,
+		Exchange:   ku.Name,
+		StartTime:  resp.Candles.StartTime,
+		Interval:   intervalString,
+		OpenPrice:  resp.Candles.OpenPrice,
+		ClosePrice: resp.Candles.ClosePrice,
+		HighPrice:  resp.Candles.HighPrice,
+		LowPrice:   resp.Candles.LowPrice,
+		Volume:     resp.Candles.TransactionVolume,
+	}
+	ku.Websocket.DataHandler <- candlestickData
+	candlestickData.AssetType = asset.Margin
+	ku.Websocket.DataHandler <- candlestickData
+	return nil
+}
+
+func (ku *Kucoin) processOrderbookWithDepth(respData []byte, instrument string) error {
+	response := WsLevel2Orderbook{}
+	err := json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	asks := make([][3]string, len(response.Asks))
+	for x := range response.Asks {
+		asks[x][0] = response.Asks[x][0]
+		asks[x][1] = response.Asks[x][1]
+	}
+	bids := make([][3]string, len(response.Bids))
+	for x := range response.Bids {
+		bids[x][0] = response.Bids[x][0]
+		bids[x][1] = response.Bids[x][1]
+	}
+	init, err := ku.UpdateLocalBuffer(&WsOrderbook{
+		Changes: OrderbookChanges{
+			Asks: asks,
+			Bids: bids,
+		},
+		Symbol:      instrument,
+		TimeMS:      response.TimeMS,
+		SequenceEnd: int64(response.TimeMS),
+	}, asset.Spot)
+	if err != nil {
+		if init {
+			return nil
+		}
+		return fmt.Errorf("%v - UpdateLocalCache for asset type: %v error: %s",
+			ku.Name,
+			asset.Spot,
+			err)
+	}
+	init, err = ku.UpdateLocalBuffer(&WsOrderbook{
+		Changes: OrderbookChanges{
+			Asks: asks,
+			Bids: bids,
+		},
+		Symbol:      instrument,
+		TimeMS:      response.TimeMS,
+		SequenceEnd: int64(response.TimeMS),
+	}, asset.Margin)
```

Comment:
And the reason I have been bringing it up is the following:
```
[ERROR] | WEBSOCKET | 28/04/2023 11:37:12 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:12 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.40000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.03648438
[ERROR] | WEBSOCKET | 28/04/2023 11:37:12 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:12 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.40000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.03648438
[ERROR] | WEBSOCKET | 28/04/2023 11:37:12 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:12 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.40000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.03648438
[ERROR] | WEBSOCKET | 28/04/2023 11:37:12 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.40000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.03648438
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00141541
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00141541
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00141541
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00141541
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00226131
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00226131
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00226131
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:13 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00226131
[ERROR] | WEBSOCKET | 28/04/2023 11:37:13 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
[INFO]  | SYNC | 28/04/2023 11:37:14 | Kucoin websocket BTC-USDT SPOT TICKER: Last 29483.30000000 Ask 29483.40000000 Bid 29483.30000000 High 0.00000000 Low 0.00000000 Volume 0.00700441
[ERROR] | WEBSOCKET | 28/04/2023 11:37:14 | could not sync new data for Kucoin BTC-USDT margin
```

This was what I meant by 
> That there are no errors that pop up when running the application
> You dont rely on a currency being enabled on spot, to populate margin